crates/cheatcodes/src/impls/mod.rs

```diff
@@ -0,0 +1,213 @@
+
+    /// Recorded storage reads and writes
+    // pub accesses: Option<RecordAccess>,
+
+    /// Recorded logs
+    // pub recorded_logs: Option<RecordedLogs>,
+
+    /// Mocked calls
+    // pub mocked_calls: BTreeMap<Address, BTreeMap<MockCallDataContext, MockCallReturnData>>,
+
+    /// Expected calls
+    // pub expected_calls: ExpectedCallTracker,
+
+    /// Expected emits
+    // pub expected_emits: VecDeque<ExpectedEmit>,
+
+    /// Map of context depths to memory offset ranges that may be written to within the call depth.
+    pub allowed_mem_writes: BTreeMap<u64, Vec<Range<u64>>>,
+
+    /// Current broadcasting information
+    // pub broadcast: Option<Broadcast>,
+
+    /// Used to correct the nonce of --sender after the initiating call. For more, check
+    /// `docs/scripting`.
+    pub corrected_nonce: bool,
+
+    /// Scripting based transactions
+    // pub broadcastable_transactions: BroadcastableTransactions,
+
+    /// Additional, user configurable context this Inspector has access to when inspecting a call
+    pub config: Arc<CheatsConfig>,
+
+    /// Test-scoped context holding data that needs to be reset every test run
+    pub context: Context,
+
+    /// Commit FS changes such as file creations, writes and deletes.
+    /// Used to prevent duplicate changes file executing non-committing calls.
+    pub fs_commit: bool,
+
+    /// Serialized JSON values.
+    pub serialized_jsons: BTreeMap<String, BTreeMap<String, Value>>,
+
+    /// Records all eth deals
+    pub eth_deals: Vec<DealRecord>,
+
+    /// Holds the stored gas info for when we pause gas metering. It is an `Option<Option<..>>`
+    /// because the `call` callback in an `Inspector` doesn't get access to
+    /// the `revm::Interpreter` which holds the `revm::Gas` struct that
+    /// we need to copy. So we convert it to a `Some(None)` in `apply_cheatcode`, and once we have
+    /// the interpreter, we copy the gas struct. Then each time there is an execution of an
+    /// operation, we reset the gas.
+    pub gas_metering: Option<Option<Gas>>,
+
+    /// Holds stored gas info for when we pause gas metering, and we're entering/inside
+    /// CREATE / CREATE2 frames. This is needed to make gas meter pausing work correctly when
+    /// paused and creating new contracts.
+    pub gas_metering_create: Option<Option<Gas>>,
+
+    /// Holds mapping slots info
+    // pub mapping_slots: Option<BTreeMap<Address, MappingSlots>>,
+
+    /// current program counter
+    pub pc: usize,
+    //// Breakpoints supplied by the `vm.breakpoint("<char>")` cheatcode
+    //// char -> pc
+    // pub breakpoints: Breakpoints,
+}
+
+/// Cheatcode implementation.
+pub(crate) trait Cheatcode: CheatcodeDef {
+    /// Applies this cheatcode to the given context.
+    ///
+    /// Implement this function if you don't need access to the EVM data.
+    fn apply(&self, state: &mut Cheatcodes) -> Result {
+        let _ = state;
+        unimplemented!()
+    }
```

Comment:
hmm, this function may not be possible to implement for cheatcodes that require the full context

Reply:
That's when you implement `apply_full`. I did to avoid having a generic on functions which don't make use of it, which is most of them

Reply:
I see, I think that's reasonable for simplifcity